### PR TITLE
*: Refine interfaces for TiFlash when using TiKV as a component

### LINKED ISSUE #12849: Support TiKV using different KvEngines
## Development Task

We need to support TiKV using a KvEngine other than engine_rocks.
1. Add some new coprosessors as complement
2. Add new features to trait engine

#### Observe execution
In AdminObserver and QueryObserver:

1. `pre_exec_admin`/`pre_exec_query` #12868
    Not all KvEngines can handle all of the commands. For example, if a KvEngine is not based on RocksDB, it may not be able to handle ComputeHash/VerifyHash. This observer allows pre-handling these commands, and filtering non-executable commands by returning false. 
3. `post_exec_admin`/`post_exec_query` #12850
    Informs results and modifications(to states) of the command, immediately after the command is executed. It returns true if one of the observes think a persistence is needed to our specific KvEngine, and we will call commit to do such persistence.
4. on_empty_cmd #12851
    Empty commands will be proposed when a new leader starts working, or processing a read-index request during leadership changing. We need to observe this, to ensure `KvEngine` can always has the up-to-date apply state. 
5. Support delay deletion of IngestSst #13061
    Some KvEngine can't actually ingest sst, since they do not use RocksDB as a backend. So they may decode the sst, and store them in memory. If we delete the sst file immediately, we may lost data if we have not persist our data in KvEngine.
    However, we can't do a persistence for every ingest sst request, this just cost too much. So we introduce this PR to allow a delay deletion of sst files, controlled by post_exec.

#### Observer applying snapshots
In ApplySnapshotObserver:
1. `pre_apply_snapshot` #12889
    It is called immediately when RegionRunner accepts one Task::Apply task.
    Since there are already `pre_apply` observer for commands, a `pre_apply` observer for applying snapshots is reasonable. We can do some statistics, transform snapshot files, batch snapshots by this observer, mostly in background threads.
2. `post_apply_snapshot` also #12889
    It is called when the snapshot is applied.
    Currently, there are only observers to inform the after when we handle a cf file. We need another observer when the whole snapshot is applied.

#### Observe RegionChange
1. `pre_commit` #12957
    Before commit and finish_for, if the observer decides it is not a proper time to do actual persist, it can skip this persistence.
    This observer is aim to disable persistence of advanced `applied_index`, since some KvEngine may not persist replicated data when `commit` and `finish_for`, so we can persist advanced `applied_index` either. 
    This can be easily done if we just disable `write_apply_state`. However, TiKV assert that data and meta should be persist as a whole, so if we don't want to persist advanced `applied_index`, we should not persist data(which is actually empty, except some pending deleting ssts) either.
    We don't worry too much about rubbish data. Since commit can also be triggered by `post_exec`, or at least `ApplyContext::end`. We don't call `write_apply_sate` in `ApplyContext::end`.

### trait engine Part
1. Introduce `can_apply_snapshot` #12924
    Before we want to handle pending snapshots, we need to call this to ask the KvEngine if it is prepared to apply snapshot. KvEngines may not be prepared if it want to batch applying snapshots. Meanwhile, ongoing background tasks introduced by `pre_apply_snapshot` will block the RegionRunner thread when applying, this is another situation we use `can_apply_snapshot`.

### PD Part
1. Enable trait engine to compute its own size, if it can be deployed in multi file systems #12948

### Others
1. We need to parameterize debug service #12960
1. We need to pub some functions/fields
2. We need to support default metapb::Store #13192 #13234
3. Bugfixs
    #13010
    #13158

### PR DESCRIPTION
**Pull request by CalvinNeo**
Signed-off-by: CalvinNeo <calvinneo1995@gmail.com>

### What is changed and how it works?

Issue Number: Close #12849

What's Changed:

```commit-message
Export necessary functions for TiFlash when using TiKV as a component
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by breezewish — APPROVED** [review]

File: src/storage/config_manager.rs [hunk 1]
```diff
@@ -19,20 +19,26 @@ use crate::{
     storage::{lock_manager::LockManager, txn::flow_controller::FlowController, TxnScheduler},
 };
 
-pub struct StorageConfigManger<E: Engine, L: LockManager> {
-    tablet_factory: Arc<dyn TabletFactory<E::Local> + Send + Sync>,
+pub struct StorageConfigManger<E: Engine, EL: engine_traits::KvEngine, L: LockManager> {
```
Comment (BusyJay) [inline src/storage/config_manager.rs:22]: Why?

Comment (CalvinNeo) [reply src/storage/config_manager.rs:22]: There are two arguments in this structure:

1. `tablet_factory` accepts an instance of `KvEngineFactory` which impls `TabletFactory<engine_rocks::Rocks>,` so `E::Local` should be `engine_rocks::RocksEngine`.
1. `scheduler` is in `Storage` which has type parameter `EK: KvEngine` which is `engine_tiflash::RocksEngine` in our case. So `E::Local` should be `engine_tiflash::RocksEngine`.

So there is a conflict here, so I want to decouple these constraints.

File: src/storage/config_manager.rs [hunk 2]
```diff
@@ -19,20 +19,26 @@ use crate::{
...
+    for StorageConfigManger<E, EL, L>
+{
+}
+unsafe impl<E: Engine, EL: engine_traits::KvEngine, L: LockManager> Sync
+    for StorageConfigManger<E, EL, L>
+{
+}
 
-impl<E: Engine, L: LockManager> StorageConfigManger<E, L> {
+impl<E: Engine, EL: engine_traits::KvEngine, L: LockManager> StorageConfigManger<E, EL, L> {
```
Comment (5kbpers) [inline src/storage/config_manager.rs:39]: Better to import `KvEngine` instead of using full path.

**Review by BusyJay — APPROVED** [review]

**Pull request merged**